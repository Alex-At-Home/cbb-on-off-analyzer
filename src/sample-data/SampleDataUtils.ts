import _ from "lodash";

export class SampleDataUtils {

  /** For snapshot testing, provides a more useful output than a giant list of stats */
  static summarizeEnrichedApiResponse(exampleRawResponse: Record<string, any>, mutableSample: boolean = false) {
    const expectedFields1 = mutableSample ? null : new Set(_.keys(exampleRawResponse));

    return {
      test: (val: any) => val && (val.off_efg ||
          val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg") ||
          val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")
        ),
        //(pick a field that all stat sets have - plus special case for: diagnostics)
      print: (val: any, serialize: ((val: any) => string), indent: ((val: any) => string)) => {
        const expectedFields = expectedFields1 || new Set(_.keys(exampleRawResponse));
        if (val.off_efg) {
          return "{\n" +
            _.chain(val).toPairs().filter(kv => {
              return !expectedFields.has(kv[0]) //derived field)
                || (!kv[1]?.hasOwnProperty("value") && exampleRawResponse[kv[0]]?.hasOwnProperty("value")) //missing value
                || (kv[1]?.hasOwnProperty("value") && _.keys(kv[1]).length > 1) //(has extra properties)
            }).sortBy(0).map(kv => {
              return indent(`${kv[0]}: ` + serialize(kv[1]))
            }).value().join("\n\n") +
          "}";
        } else if (val.hasOwnProperty("dRtg") || val.hasOwnProperty("oRtg")) {
          return JSON.stringify(_.pick(val, [
            "oRtg", "adjORtg", "adjORtgPlus", "dRtg", "adjDRtg", "adjDRtgPlus"
          ]), null , 3);
        } else if (val.hasOwnProperty("deltaOffPpp") || val.hasOwnProperty("deltaDefPpp")) {
          return JSON.stringify(_.pick(val, [
            "deltaOffPpp", "deltaOffEfg", "deltaOffAdjEff", "deltaDefPpp", "deltaDefEfg", "deltaDefAdjEff"
          ]), null , 3);
        } else {
          return "???";
        }
      }
    }
  }


  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateLineup(input: Record<string, any>) {
    const lineupAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.toPairs(lineupAsObj).map(kv => {
      return { key: kv[0], ...(kv[1] as Record<string, any>) };
    })
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplateTeam(input: Record<string, any>) {
    const teamAsObj = _.transform(_.toPairs(input), (acc, inElKv) => {
      _.forEach(_.toPairs(inElKv[1] as Record<string, any>), (kv) => {
        _.forEach(_.toPairs(kv[1] as Record<string, any>), (deepKv) => {
          if (inElKv[0] != "main") {
            _.set(acc, `${kv[0]}.${deepKv[0]}${inElKv[0]}.value`, deepKv[1]);
          } else {
            _.set(acc, `${deepKv[0]}.${kv[0]}`, deepKv[1]);
          }
        });
      });
    }, {} as Record<string, any>);
    return teamAsObj;
  }

  /** From easier to read templates builds a bucket list */
  static buildResponseFromTemplatePlayer(input: Record<string, any>) {
    const flatPlayersAsObj = SampleDataUtils.buildResponseFromTemplateTeam(input);
    return _.transform(["baseline", "on", "off"], (acc, v) => {
      acc[v] = { player: {
        buckets: _.chain(flatPlayersAsObj).toPairs().filter(kv => {
          return kv[0].indexOf(v) == 0;
        }).map(kv => {
          const key = kv[0].substring(v.length + 2); //(+2 for ": ")
          return { ...(kv[1]), key: key }
        }).value()
      } };
    }, {} as Record<string, any>);
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseTeam(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().map((kv: [string, any]) => {
        return { key: kv[0], ...(kv[1]) }
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponsePlayer(input: Record<string, Record<string, any>>) {
    return SampleDataUtils.buildTemplateFromResponseLineup(
      _.chain(input).toPairs().flatMap((kv: [string, any]) => {
        const players = (kv[1]?.player?.buckets || []) as Array<Record<string, any>>;
        return players.map(p => { return { ...p, key: `${kv[0]}: ${p.key}` } });
      }).value()
    );
  }

  /** Builds an easier to manipulate template for the different responses, from an actual response */
  static buildTemplateFromResponseLineup(input: Array<Record<string, any>>, filter?: Set<String>) {
    const preSorted = _.transform(input, (acc, inEl) => {
      const recordId = inEl?.key || "unknown";
      const injectInfo = (k: string, prefix: string) => {
        const topLevelKey = k.substring(prefix.length);
        const realPrefix = prefix || "main";
        if (prefix) {
          const existingObj = _.get(acc, `${topLevelKey}.${recordId}`) || {};
          existingObj[realPrefix] = inEl[k]?.value;
          const sortedObj = _(existingObj).toPairs().sortBy(0).reverse().fromPairs().value()
          _.set(acc, `${topLevelKey}.${recordId}`, sortedObj);
        } else {
          _.set(acc, `main.${topLevelKey}.${recordId}`, inEl[k]);
        }
      };

      if (!filter || filter.has(recordId)) {
        _.keys(inEl).filter(k => k != "key").forEach(k => {
          const stat = _.find([
            "oppo_total_off_", "oppo_total_def_",
            "oppo_off_", "oppo_def_",

            "team_total_off_", "team_total_def_",
            "team_off_", "team_def_",

            "total_off_", "total_def_",
            "off_", "def_"
          ], prefix => k.indexOf(prefix) == 0);
          if (stat) {
            injectInfo(k, stat);
          } else {
            injectInfo(k, "");
          }
        });
      }
    }, {} as Record<string, any>);

    // Sort the keys into some sort of logical order
    return _.chain(preSorted).toPairs().sortBy([
      (kv: [ string , any ]) => {
        const key = kv[0];
        if (key == "main") {
          return 100;
        } else if ((key.indexOf("assist") >= 0) || (key.indexOf("to") >= 0) || (key.indexOf("ast") >= 0)) {
          return 4;
        } else if ((key.indexOf("ft") == 0) || (key.indexOf("fg") == 0) || (key == "efg")) {
          return 1;
        } else if ((key.indexOf("2p") == 0) || (key.indexOf("3p") == 0)) {
          return 2;
        } else if ((key.indexOf("rb") > 0)) {
          return 3;
        } else if ((key.indexOf("adj") >= 0)) {
          return 5;
        } else { //misc other stuff
          return 6;
        }
      },
      (kv: [ string , any ]) => {
        return kv[0];
      },
    ]).fromPairs().value();
  }
}
