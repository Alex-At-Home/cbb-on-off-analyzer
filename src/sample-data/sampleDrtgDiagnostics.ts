export const sampleDrtgDiagnostics = {
  "stl": 30,
  "blk": 5,
  "drb": 94,
  "pfPct": 0.2,
  "playerRtg": 142.82324406037372,
  "playerDelta": 8.134297736197055,
  "scPossConceded": 0.6341530662717683,
  "noShotCredit": 32.188268933121634,
  "reboundCredit": 36.07039815349106,
  "missFtCredit": 6.070287539936104,
  "stopsIndPct": 0.21044437889736348,
  "stopsTeamPct": 0.15540255483086815,
  "teamBlk": 60,
  "oppoPts": 1804,
  "oppoPoss": 1766,
  "oppoFga": 1429,
  "oppoFgm": 534,
  "oppoFtm": 426,
  "oppoFta": 626,
  "oppoFtPoss": 297.34999999999997,
  "oppoTov": 59,
  "teamStl": 150,
  "opponentOrbPct": 0.2708757637474542,
  "opponentFgPct": 0.3736878936319104,
  "teamOrbCreditToDefender": 0.2724649000411893,
  "teamOrbCreditToRebounder": 0.16965277015673305,
  "teamDvsRebCredit": 0.616272360069244,
  "oppoFgMiss": 835,
  "oppoNonStlTov": -91,
  "teamMissWeight": 0.4376537866243272,
  "oppoFtPct": 0.6805111821086262,
  "oppoFtHitOnePlus": 266.99856230031946,
  "oppoProbFtHitOnePlus": 0.8979268952423726,
  "oppoScPoss": 800.9985623003195,
  "oppoPtsPerScore": 2.2521888114496065,
  "teamRtg": 102.15175537938845,
  "adjDRtg": 102.04282992219363,
  "adjDRtgPlus": 0.4085659844387266,
  "avgEff": 100,
  "dRtg": 110.2860531155855,
  "offSos": 108.0781993204983
 };
