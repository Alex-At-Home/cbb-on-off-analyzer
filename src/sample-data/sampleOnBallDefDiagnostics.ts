/** Manually built by copying the snapshot from src/utils/stats/__tests__/RatingUtils.test.ts */
export const sampleOnBallDefDiagnostics =
{
   "playerRebCredit": 0.09664054373332856,
   "teamRebCredit": 0.14616882239665946,
   "comboRebCredit": 0.2948067270155614,
   "playerTargetPoss": 185.586,
   "targetedPct": 0.12134831460674157,
   "playerPtsPerScore": 2.4291266575217194,
   "onBallStopCredit": 0.4668039520787198,
   "offBallPoss": 1572.7865168539327,
   "offBallPts": 1539.0286760055637,
   "otherRebounds": 631.6805699093197,
   "otherRebCredit": 0.14474344317512738,
   "offBallStopCredit": 0.39516433361719344,
   "onVsOffBallWeight": 0.4,
   "weightedPtsPerScore": 2.126826397164934,
   "onBallCreditWeight": 0.04853932584269663,
   "offBallCreditWeight": 0.13179775280898876,
   "comboBallStopCredit": 0.8290444280732188,
   "unadjDRtg": 103.9658869175629,
   "weightedClassicDRtgMean": 98.84955759282161,
   "weightedUnadjDRtgMean": 107.19011038664136,
   "uncategorizedAdjustment": -7.10724422350558,
   "adjustedPossPct": 0.08000000000000002,
   "dRtg": 96.85864269405732,
   "adjDRtg": 89.6509562344713,
   "adjDRtgPlus": -2.069808753105741
};
