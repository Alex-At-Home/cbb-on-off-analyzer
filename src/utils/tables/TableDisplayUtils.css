.assistBadge::after {
  content: "A"
}
.megaAssistBadge::after {
  content: "A";
  font-weight: 700;
}
.freeThrowBadge::after {
  content: "F"
}
.megaFreeThrowBadge::after {
  content: "F";
  font-weight: 700;
}
.twoPointBadge::after {
  content: "2";
}
.megaTwoPointBadge::after {
  content: "2";
  font-weight: 700;
}
.threePointBadge::after {
  content: "3";
}
.megaThreePointBadge::after {
  content: "3";
  font-weight: 700;
}
