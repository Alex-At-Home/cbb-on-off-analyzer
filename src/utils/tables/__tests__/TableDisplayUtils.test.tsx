import renderer from 'react-test-renderer';
import React from 'react';
import { TableDisplayUtils } from '../TableDisplayUtils';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import _ from "lodash";

import { SampleDataUtils } from "../../../sample-data/SampleDataUtils";
import { sampleLineupStatsResponse } from "../../../sample-data/sampleLineupStatsResponse";
import { samplePlayerStatsResponse } from "../../../sample-data/samplePlayerStatsResponse";
import { sampleTeamStatsResponse } from "../../../sample-data/sampleTeamStatsResponse";

describe("TableDisplayUtils", () => {

  const testLineup = [
    { code: "p1", id: "player1" },
    { code: "p2", id: "player2" },
    { code: "p3", id: "player3" },
    { code: "p4", id: "player4" },
    { code: "p5", id: "player5" },
  ];
  const positionFromPlayerKey = { //(just used for display)
    "player1": { posClass: "p-PG" },
    "player2": { posClass: "CG" },
    "player3": { posClass: "WG" },
    "player4": { posClass: "WF" },
    "player5": { posClass: "PF/C" },
  };
  const perLineupPlayerMap = {
    "player1": { // *3* A
        off_rtg: { value: 101 }, off_adj_rtg: { value : 0.5 }, //(rtg just used for display)
        off_usage: { value: 0.1 }, off_assist: { value: 0.20 },
        off_3pr: { value: 0.8 }, off_ftr: { value: 0.1 },
        def_orb: { value: 0.05 }, //(just used for display)
    },
    "player2": { // *A* *F*
        off_rtg: { value: 102 }, off_adj_rtg: { value : 1.0 }, //(rtg just used for display)
        off_usage: { value: 0.15 }, off_assist: { value: 0.30 },
        off_3pr: { value: 0.3 }, off_ftr: { value: 0.9 },
        def_orb: { value: 0.10 }, //(just used for display)
    },
    "player3": { // *2*
        off_rtg: { value: 103 }, off_adj_rtg: { value : 1.5 }, //(rtg just used for display)
        off_usage: { value: 0.20 }, off_assist: { value: 0.15 },
        off_3pr: { value: 0.01 }, off_ftr: { value: 0.3 },
        def_orb: { value: 0.15 }, //(just used for display)
    },
    "player4": { // 2 F
        off_rtg: { value: 104 }, off_adj_rtg: { value : 2.0 }, //(rtg just used for display)
        off_usage: { value: 0.25 }, off_assist: { value: 0.15 },
        off_3pr: { value: 0.15 }, off_ftr: { value: 0.5 },
        def_orb: { value: 0.20 }, //(just used for display)
    },
    "player5": { // 3 A F
        off_rtg: { value: 105 }, off_adj_rtg: { value : 2.5 }, //(rtg just used for display)
        off_usage: { value: 0.30 }, off_assist: { value: 0.20 },
        off_3pr: { value: 0.50 }, off_ftr: { value: 0.5 },
        def_orb: { value: 0.25 }, //(just used for display)
    },
  }

  test("buildDecoratedLineup - plain version", () => {
    const component = renderer.create(
      <span>{
        TableDisplayUtils.buildDecoratedLineup(
          "p1_p2_p3_p4_p5", testLineup, perLineupPlayerMap, positionFromPlayerKey, "off_adj_rtg", false
        )
      }</span>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("buildDecoratedLineup - decorated version", () => {
    const component = renderer.create(
      <span>{
        TableDisplayUtils.buildDecoratedLineup(
          "p1_p2_p3_p4_p5", testLineup, perLineupPlayerMap, positionFromPlayerKey, "off_adj_rtg", true
        )
      }</span>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("buildDecoratedLineup - buildTooltipTexts", () => {
    const component = renderer.create(
      <span>{
        TableDisplayUtils.buildTooltipTexts(
          "p1_p2_p3_p4_p5", testLineup, perLineupPlayerMap, positionFromPlayerKey
        )
      }</span>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  // Mutate expect here for better display:

  const sampleData = {} as Record<string, any>;
  const injectData = (mutableIn: Record<string, any>, newSample: Record<string, any>) => {
    const fields = _.keys(mutableIn);
    fields.forEach(key => delete mutableIn[key]);
    _.toPairs(newSample).forEach(kv => mutableIn[kv[0]] = kv[1]);
  }

  expect.addSnapshotSerializer(SampleDataUtils.summarizeEnrichedApiResponse(
    sampleData, true
  ));

  test("injectPlayTypeInfo - lineups", () => {
    injectData(sampleData, sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets[1]);
    expect(TableDisplayUtils.injectPlayTypeInfo(
      sampleLineupStatsResponse.responses[0].aggregations.lineups.buckets[0],
      false, false
    )).toMatchSnapshot();
  });

  test("injectPlayTypeInfo - players (!expanded)", () => {
    injectData(sampleData,
      samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets[1]
    );
    expect(TableDisplayUtils.injectPlayTypeInfo(
      samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets[0],
      false, true
    )).toMatchSnapshot();
  });

  test("injectPlayTypeInfo - players (expanded)", () => {
    injectData(sampleData,
      samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets[1]
    );
    expect(TableDisplayUtils.injectPlayTypeInfo(
      samplePlayerStatsResponse.responses[0].aggregations.tri_filter.buckets.baseline.player.buckets[0],
      true, true
    )).toMatchSnapshot();
  });
});
