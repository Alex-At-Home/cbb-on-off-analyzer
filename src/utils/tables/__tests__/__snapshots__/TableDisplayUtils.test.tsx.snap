// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableDisplayUtils addQueryInfo 1`] = `<div />`;

exports[`TableDisplayUtils addQueryInfo 2`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="baseline0QueryInfo"
      placement="right"
    >
      query: 'BQ', filters: 'QF' ([100] team possessions)
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 3`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="on0QueryInfo"
      placement="right"
    >
      BASE:[query: 'BQ', filters: 'QF'] ([100] team possessions)
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 4`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="on0QueryInfo"
      placement="right"
    >
      query: 'OQ' ([100] team possessions)
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 5`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="on0QueryInfo"
      placement="right"
    >
      filters: 'OF' AND BASE:[query: 'BQ']
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 6`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="off0QueryInfo"
      placement="right"
    >
      query: 'OffQ', filters: 'OffF' ([200] team possessions)
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 7`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="off0QueryInfo"
      placement="right"
    >
      query: 'OffQ' AND BASE:[filters: 'QF']
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 8`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="other0QueryInfo"
      placement="right"
    >
      query: 'OffQ', filters: 'OffF' ([200] team possessions)
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils addQueryInfo 9`] = `
<OverlayTrigger
  overlay={
    <Tooltip
      id="other0QueryInfo"
      placement="right"
    >
      query: 'OffQ' AND BASE:[filters: 'QF']
    </Tooltip>
  }
  placement="auto"
>
  <u
    style={
      Object {
        "textDecorationStyle": "dotted",
      }
    }
  >
    <div />
  </u>
</OverlayTrigger>
`;

exports[`TableDisplayUtils buildDecoratedLineup - buildTooltipTexts 1`] = `
<span>
  <div
    className="tooltip bs-tooltip-right"
    id="p1_p2_p3_p4_p5_info"
    role="tooltip"
    x-placement="right"
  >
    <div
      className="arrow"
    />
    <div
      className="tooltip-inner"
    >
      <span>
        player1: p-PG
        <br />
      </span>
      <span>
        ORtg 101 on 10%, DRB 5%
        <br />
      </span>
      <span>
        
        <br />
      </span>
      <span>
        player2: CG
        <br />
      </span>
      <span>
        ORtg 102 on 15%, DRB 10%
        <br />
      </span>
      <span>
        
        <br />
      </span>
      <span>
        player3: WG
        <br />
      </span>
      <span>
        ORtg 103 on 20%, DRB 15%
        <br />
      </span>
      <span>
        
        <br />
      </span>
      <span>
        player4: WF
        <br />
      </span>
      <span>
        ORtg 104 on 25%, DRB 20%
        <br />
      </span>
      <span>
        
        <br />
      </span>
      <span>
        player5: PF/C
        <br />
      </span>
      <span>
        ORtg 105 on 30%, DRB 25%
        <br />
      </span>
      <span>
        
        <br />
      </span>
      Click to open a more detailed analysis page for this lineup 
    </div>
  </div>
</span>
`;

exports[`TableDisplayUtils buildDecoratedLineup - decorated version 1`] = `
<span>
  <div>
    <a
      href="/OnOffAnalyzer?autoOffQuery=false&calcRapm=true&filter=p1%2Cp2%2Cp3%2Cp4%2Cp5&offQuery=&onQuery=%7B%22player1%22%3B%22player2%22%3B%22player3%22%3B%22player4%22%3B%22player5%22%7D%3D5&possAsPct=false&showExpanded=true&showExtraInfo=true&showRoster=true&showTeamPlayTypes=true&teamShotCharts=true&"
      onClick={[Function]}
      style={
        Object {
          "color": "inherit",
          "textDecoration": "inherit",
        }
      }
      target="_blank"
    >
      <span>
        <span
          style={
            Object {
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="badge badge-light"
            style={
              Object {
                "backgroundColor": "#f2f9f280",
              }
            }
          >
            <span
              style={
                Object {
                  "fontSize": "small",
                  "fontWeight": 100,
                }
              }
            >
              p1
            </span>
          </span>
          <span
            style={Object {}}
          >
            <sup
              className="megaThreePointBadge"
            />
            <sup
              className="assistBadge"
            />
          </span>
        </span>
        <span
          style={
            Object {
              "opacity": 0,
            }
          }
        >
           ; 
        </span>
      </span>
      <span>
        <span
          style={
            Object {
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="badge badge-light"
            style={
              Object {
                "backgroundColor": "#e5f2e580",
              }
            }
          >
            <span
              style={
                Object {
                  "fontSize": "small",
                  "fontWeight": 300,
                }
              }
            >
              p2
            </span>
          </span>
          <span
            style={Object {}}
          >
            <sub
              className="megaFreeThrowBadge"
            />
            <sup
              className="megaAssistBadge"
            />
          </span>
        </span>
        <span
          style={
            Object {
              "opacity": 0,
            }
          }
        >
           ; 
        </span>
      </span>
      <span>
        <span
          style={
            Object {
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="badge badge-light"
            style={
              Object {
                "backgroundColor": "#d9ecd980",
              }
            }
          >
            <span
              style={
                Object {
                  "fontSize": "small",
                  "fontWeight": 500,
                }
              }
            >
              p3
            </span>
          </span>
          <span
            style={Object {}}
          >
            <sup
              className="megaTwoPointBadge"
            />
          </span>
        </span>
        <span
          style={
            Object {
              "opacity": 0,
            }
          }
        >
           ; 
        </span>
      </span>
      <span>
        <span
          style={
            Object {
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="badge badge-light"
            style={
              Object {
                "backgroundColor": "#cce6cc80",
              }
            }
          >
            <span
              style={
                Object {
                  "fontSize": "small",
                  "fontWeight": 600,
                }
              }
            >
              p4
            </span>
          </span>
          <span
            style={Object {}}
          >
            <sup
              className="twoPointBadge"
            />
            <sub
              className="freeThrowBadge"
            />
          </span>
        </span>
        <span
          style={
            Object {
              "opacity": 0,
            }
          }
        >
           ; 
        </span>
      </span>
      <span>
        <span
          style={
            Object {
              "whiteSpace": "nowrap",
            }
          }
        >
          <span
            className="badge badge-light"
            style={
              Object {
                "backgroundColor": "#bfdfbf80",
              }
            }
          >
            <span
              style={
                Object {
                  "fontSize": "small",
                  "fontWeight": 700,
                }
              }
            >
              p5
            </span>
          </span>
          <span
            style={Object {}}
          >
            <sup
              className="threePointBadge"
            />
            <sub
              className="freeThrowBadge"
            />
            <sup
              className="assistBadge"
            />
          </span>
        </span>
      </span>
    </a>
  </div>
</span>
`;

exports[`TableDisplayUtils buildDecoratedLineup - plain version 1`] = `
<span>
  <div>
    <span>
      <a
        href="#"
        style={
          Object {
            "color": "inherit",
            "textDecoration": "inherit",
          }
        }
      >
        <b>
          p1 / p2 / p3 / p4 / p5
        </b>
      </a>
    </span>
  </div>
</span>
`;

exports[`TableDisplayUtils injectPlayTypeInfo - lineups 1`] = `
{
  def_2pmidr: Object {
    "extraInfo": <span>
      23% assisted
    </span>,
    "value": 0.2997032640949555,
  }

  def_2prim: Object {
    "extraInfo": <span>
      Block:
       
      11.8
      %
    </span>,
    "value": 0.5436893203883495,
  }

  def_2primr: Object {
    "extraInfo": <span>
      45% assisted
    </span>,
    "value": 0.3056379821958457,
  }

  def_3pr: Object {
    "extraInfo": <span>
      88% assisted
    </span>,
    "value": 0.39465875370919884,
  }

  def_assist: Object {
    "extraInfo": <span>
      Assist breakdown:
      <li>
        52
        % for 3P
      </li>
      <li>
        10
        % for mid-range
      </li>
      <li>
        37
        % at the rim
      </li>
    </span>,
    "value": 0.5275590551181102,
  }

  def_ftr: Object {
    "extraInfo": <span>
      FT: 
      74.7
      %
    </span>,
    "value": 0.27002967359050445,
  }

  def_ppp: Object {
    "extraInfo": <span>
      Play category breakdown:
      <li>
        20.6
        % transition:
        <br />
        +
        6.9
         pts/100
      </li>
      <li>
        8.3
        % scramble:
        <br />
        +
        9.3
         pts/100
      </li>
    </span>,
    "value": 93.29896907216495,
  }

  def_to: Object {
    "extraInfo": <span>
      Stl: 
      8.0
      %
      <br />
      Non-Stl TO:
       
      7.2
      %
    </span>,
    "value": 0.15206185567010308,
  }

  off_2pmidr: Object {
    "extraInfo": <span>
      21% assisted
    </span>,
    "value": 0.17956656346749225,
  }

  off_2primr: Object {
    "extraInfo": <span>
      43% assisted
    </span>,
    "value": 0.3653250773993808,
  }

  off_3pr: Object {
    "extraInfo": <span>
      78% assisted
    </span>,
    "value": 0.4551083591331269,
  }

  off_assist: Object {
    "extraInfo": <span>
      Assist breakdown:
      <li>
        47
        % for 3P
      </li>
      <li>
        4
        % for mid-range
      </li>
      <li>
        49
        % at the rim
      </li>
    </span>,
    "value": 0.5211267605633803,
  }

  off_ftr: Object {
    "extraInfo": <span>
      FT: 
      76.6
      %
    </span>,
    "value": 0.5696594427244582,
  }

  off_poss: Object {
    "extraInfo": <span>
      71.7
       poss/g
    </span>,
    "value": 401,
  }

  off_ppp: Object {
    "extraInfo": <span>
      Play category breakdown:
      <li>
        25.3
        % transition:
        <br />
        +
        16.6
         pts/100
      </li>
      <li>
        9.6
        % scramble:
        <br />
        +
        8.1
         pts/100
      </li>
    </span>,
    "value": 117.2069825436409,
  }}
`;

exports[`TableDisplayUtils injectPlayTypeInfo - players (!expanded) 1`] = `
{
  def_2pmidr: <small
    style={
      Object {
        "textShadow": "0px 0px 15px #add8e6,0px 0px 15px #add8e6,0px 0px 15px #add8e6",
      }
    }
  >
    <i>
      0
      %
    </i>
  </small>

  def_2primr: <small
    style={
      Object {
        "textShadow": "0px 0px 15px #e0f0f5,0px 0px 15px #e0f0f5,0px 0px 15px #e0f0f5",
      }
    }
  >
    <i>
      26
      %
    </i>
  </small>

  def_3pr: <small
    style={
      Object {
        "textShadow": "0px 0px 15px #ffa500,0px 0px 15px #ffa500,0px 0px 15px #ffa500",
      }
    }
  >
    <i>
      63
      %
    </i>
  </small>

  off_2pmidr: Object {
    "extraInfo": <span>
      0% assisted
    </span>,
    "value": 0.18258426966292135,
  }

  off_2primr: Object {
    "extraInfo": <span>
      26% assisted
    </span>,
    "value": 0.32865168539325845,
  }

  off_3pr: Object {
    "extraInfo": <span>
      63% assisted
    </span>,
    "value": 0.4887640449438202,
  }

  off_assist: Object {
    "extraInfo": <span>
      Assist breakdown:
      <li>
        48
        % for 3P
      </li>
      <li>
        7
        % for mid-range
      </li>
      <li>
        46
        % at the rim
      </li>
    </span>,
    "value": 0.2916666666666667,
  }

  off_ftr: Object {
    "extraInfo": <span>
      FT: 
      81.1
      %
    </span>,
    "value": 0.5955056179775281,
  }

  off_team_poss: Object {
    "extraInfo": <span>
      80.1
       poss/g
    </span>,
    "value": 1802,
  }}
`;

exports[`TableDisplayUtils injectPlayTypeInfo - players (expanded) 1`] = `
{
  def_2pmidr: <small
    style={
      Object {
        "textShadow": "0px 0px 15px #add8e6,0px 0px 15px #add8e6,0px 0px 15px #add8e6",
      }
    }
  >
    <i>
      0
      %
    </i>
  </small>

  def_2primr: <small
    style={
      Object {
        "textShadow": "0px 0px 15px #e0f0f5,0px 0px 15px #e0f0f5,0px 0px 15px #e0f0f5",
      }
    }
  >
    <i>
      26
      %
    </i>
  </small>

  def_3pr: <small
    style={
      Object {
        "textShadow": "0px 0px 15px #ffa500,0px 0px 15px #ffa500,0px 0px 15px #ffa500",
      }
    }
  >
    <i>
      63
      %
    </i>
  </small>

  off_assist: Object {
    "extraInfo": <span>
      Assist breakdown:
      <li>
        48
        % for 3P
      </li>
      <li>
        7
        % for mid-range
      </li>
      <li>
        46
        % at the rim
      </li>
    </span>,
    "value": 0.2916666666666667,
  }

  off_ftr: Object {
    "extraInfo": <span>
      FT: 
      81.1
      %
    </span>,
    "value": 0.5955056179775281,
  }

  off_team_poss: Object {
    "extraInfo": <span>
      80.1
       poss/g
    </span>,
    "value": 1802,
  }}
`;
